state/protocol/badger/state.go

```diff
@@ -33,17 +40,26 @@ type State struct {
 		statuses storage.EpochStatuses
 	}
 	versionBeacons storage.VersionBeacons
-	// cache the root height because it cannot change over the lifecycle of a protocol state instance
+
+	// rootHeight marks the cutoff of the history this node knows about. We cache it in the state
+	// because it cannot change over the lifecycle of a protocol state instance. It is frequently
+	// larger than the height of the root block of the spork, (also cached below as
+	// `sporkRootBlockHeight`), for instance if the node joined in an epoch after the last spork.
 	rootHeight uint64
-	// cache the spork root block height because it cannot change over the lifecycle of a protocol state instance
+	// sporkRootBlockHeight is the height of the root block in the current spork. We cache it in
+	// the state, because it cannot change over the lifecycle of a protocol state instance.
+	// Caution: A node that joined in a later epoch past the spork, the node will likely _not_
+	// know the spork's root block in full (though it will always know the height).
 	sporkRootBlockHeight uint64
+	// cache the latest finalized block
+	cachedFinal *atomic.Pointer[cachedHeader]
 }
```

Comment:
I wonder if makes sense to also add `cachedSealed`. I think will be useful for AN, EN and VN.

Reply:
Added it here: https://github.com/onflow/flow-go/pull/4145/commits/c45c3d6ab08385f5acfcf3bbaab1552316510be5